action/protocol/poll/nativestaking.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2019 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package poll
+
+import (
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-election/types"
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/action/protocol"
+	"github.com/iotexproject/iotex-core/state"
+)
+
+var (
+	dummyCaller, _ = address.FromString(address.ZeroAddress)
+
+	// ErrNoData is an error that there's no data in the contract
+	ErrNoData = errors.New("no data")
+)
+
+type (
+	// NativeStaking represents native staking struct
+	NativeStaking struct {
+		cm           protocol.ChainManager
+		getBlockTime GetBlockTime
+		contract     string
+		abi          abi.ABI
+	}
+
+	pygg struct {
+		Count           *big.Int
+		Indexes         []*big.Int
+		StakeStartTimes []*big.Int
+		StakeDurations  []*big.Int
+		Decays          []bool
+		StakedAmounts   []*big.Int
+		CanNames        [][12]byte
+		Owners          []common.Address
+	}
+
+	// VoteTally is a map of candidates on native chain
+	VoteTally map[[12]byte]*state.Candidate
+)
+
+// NewNativeStaking creates a NativeStaking instance
+func NewNativeStaking(cm protocol.ChainManager, getBlockTime GetBlockTime, staking string) (*NativeStaking, error) {
+	abi, err := abi.JSON(strings.NewReader(NsAbi))
+	if err != nil {
+		return nil, err
+	}
+	if cm == nil {
+		return nil, errors.New("failed to create native staking: empty chain manager")
+	}
+	if getBlockTime == nil {
+		return nil, errors.New("failed to create native staking: empty getBlockTime")
+	}
+	if _, err := address.FromString(staking); err != nil {
+		return nil, errors.Errorf("invalid staking contract %s", staking)
+	}
+	return &NativeStaking{cm, getBlockTime, staking, abi}, nil
+}
+
+// Votes returns the votes on height
+func (ns *NativeStaking) Votes(height uint64) (VoteTally, error) {
+	if ns.contract == "" {
+		return nil, ErrNoData
+	}
+
+	now, err := ns.getBlockTime(height)
```

Comment:
concern: the height is actually epoch height, and we use epoch height's time to calculate votes based on the buckets of current state. This may cause inconsistency in the future once we support archive state db.

Reply:
fixed in latest push